File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int N = friends.length;
        HashMap<String, Integer> idxMap = new HashMap<>();
        for (int i=0;i<N;i++) {
            idxMap.put(friends[i], i);
        }
        
        int[][] record = new int[N][N];
        int[] score = new int[N];
        for (String gift: gifts) {
            String[] temp = gift.split(" ");
            int give = idxMap.get(temp[0]);
            int take = idxMap.get(temp[1]);
            record[give][take]++;
            score[give]++;
            score[take]--;
        }
        
        int answer = 0;
        for (int i=0;i<N;i++) {
            int count = 0;
            for (int j=0;j<N;j++) {
                if (i == j) continue;
                if (record[i][j] > record[j][i] || (record[i][j] == record[j][i] && score[i] > score[j])) {
                    count++;
                }
            }
            answer = Math.max(answer, count);
            
        }
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    N = len(friends)
    friend_dic = {}
    for i in range(N):
        friend_dic[friends[i]] = i
    
    give_get = [[0] * N for _ in range(N)]
    for gift in gifts:
        give, get = gift.split()
        give_get[friend_dic[give]][friend_dic[get]] += 1
    
    get = [0] * N
    for A in range(N):
        for B in range(N):
            if give_get[A][B] > give_get[B][A]:
                get[A] += 1
            elif give_get[A][B] == give_get[B][A]:
                A_give = sum(give_get[A])
                B_give = sum(give_get[B])
                A_get, B_get = 0, 0
                for i in range(N):
                    A_get += give_get[i][A]
                    B_get += give_get[i][B]
                if A_give - A_get > B_give - B_get:
                    get[A] += 1
    
    return max(get)